src/widget/wcoverartmenu.cpp

```diff
@@ -68,22 +71,40 @@ void WCoverArtMenu::slotChange() {
         return;
     }
 
-    // TODO(rryan): Ask if user wants to copy the file.
+    QString selectedCoverExtension = QFileInfo(selectedCoverPath).suffix();
+
+    QString coverArtCopyFilePath =
+            trackFileInfo.absoluteFilePath().left(
+                    trackFileInfo.absoluteFilePath().lastIndexOf('.') + 1) +
+            selectedCoverExtension;
 
-    CoverInfoRelative coverInfo;
-    // Create a security token for the file.
-    auto selectedCover = mixxx::FileAccess(mixxx::FileInfo(selectedCoverPath));
-    QImage image(selectedCoverPath);
-    if (image.isNull()) {
-        // TODO(rryan): feedback
+    VERIFY_OR_DEBUG_ASSERT(m_isWorkerRunning == false) {
         return;
-    }
-    coverInfo.type = CoverInfo::FILE;
-    coverInfo.source = CoverInfo::USER_SELECTED;
-    coverInfo.coverLocation = selectedCoverPath;
-    coverInfo.setImage(image);
-    qDebug() << "WCoverArtMenu::slotChange emit" << coverInfo;
-    emit coverInfoSelected(coverInfo);
+    };
+
+    m_worker.reset(new CoverArtCopyWorker(selectedCoverPath, coverArtCopyFilePath));
+
+    connect(m_worker.data(),
+            &CoverArtCopyWorker::finished,
+            this,
+            &WCoverArtMenu::slotFinished);
+
+    connect(m_worker.data(),
+            &CoverArtCopyWorker::askOverwrite,
+            this,
+            &WCoverArtMenu::slotAskOverwrite);
+
+    connect(m_worker.data(),
+            &CoverArtCopyWorker::updateCoverArt,
+            this,
+            &WCoverArtMenu::slotCoverArtUpdated);
+
+    connect(m_worker.data(),
+            &CoverArtCopyWorker::started,
+            this,
+            &WCoverArtMenu::slotStarted);
+
+    m_worker->start();
```

Comment:
the waiting code in the destructor is missing.

Reply:
I couldn't understand it could you please explain it a bit more please? Thanks in advance.

Reply:
Something like this: 
https://github.com/mixxxdj/mixxx/blob/0580818bf89b2d709f222ab2d315eccc47f31ec5/src/engine/engineworkerscheduler.cpp#L19

Reply:
Oh, destructor should have this in order to not terminate the working thread, right?

Reply:
right.